### DISCUSSION TITLE
Use of HTTP Methods

### DISCUSSION THREAD

**Discussion by ethndotsh** [blitz-js/blitz#2958] [Ideas] [answered]
Currently, the external API that third-parties can access has a rather unusual style, in my opinion. E.g.

Get user
`GET  /api/getUser/[userId]`

Create user
`GET  /api/createUser?email=[email]?password=[password]`

But, rather than that, use different HTTP methods on the same endpoint. E.g.

Get a user
`GET   /api/user/[userId]`

Create a user
`POST  /api/user`

Delete a user
`DELETE /api/user/[userId]`

Update a user
`PUT   /api/user/[userId]`

That's what I believe would be better, but please correct me if I am wrong. Also please correct me if I did not showcase the current endpoints correctly.

**Comment by beerose** [comment]
The naming conventions come from the fact that Blitz exposes RPC API. 

The naming you provided is more in a REST style. You can think of it this way: 
* in REST (representational **state** transfer) you have the state, a resource, that you access, modify and so on. Hence you have `users/` and you can fetch them, delete them, etc.
* in RPC (remote **procedure** call) you have procedures, and you call the procedures. It's like literally calling a function, taking a method name and arguments. But that's done over HTTP. That's also why the endpoints are named as your queries/mutations.

**Comment by MrLeebo** [comment, marked as answer]
To tack onto @beerose's answer, if you want to craft an API suited for third parties, you can always create a separate endpoint using [API Routes](https://blitzjs.com/docs/api-routes) and follow RESTful patterns in those routes. Third parties don't benefit from the automatic code generation of queries/mutations (yet!) so they'd probably prefer to have a route tailored for them like that anyway.

**Reply by ethndotsh** [reply]
I see. I thought that all my auto-generated models and their queries/mutations were automatically used for both the data later but also the third-party API.

Or does it do this? This is a very cool framework, just a few things here and there that confuse me.

**Reply by beerose** [reply]
Yeah, queries/mutations are exposed as RPC API, and that's also what makes it possible to use them directly from the frontend code. At build time, the direct function import is replaced with an RPC network call.

Here's more on RPC API: https://blitzjs.com/docs/rpc-specification
